backend/common/utils/rollup.py

```diff
@@ -32,6 +32,56 @@ def ancestors(cell_type):
     return ancestors
 
 
+def get_valid_descendants(
+    cell_type: str, valid_cell_types: frozenset[str], cell_counts: Optional[dict[str, int]] = None
+):
+    """
+    Get valid descendants for a cell type.
+
+    Terms can be suffixed with ";;{A_0}--{B_0}--..." to indicate that only descendants within the same
+    group specified by the suffix should be considered. For example, if the term is
+    "CL:0000540;;{tissue_0}--{disease_0}" then only descendants with the same tissue_0 and disease_0
+    will be considered.
+
+    Arguments
+    ---------
+    cell_type : str
+        Cell type (cell type ontology term ID, potentially suffixed)
+    valid_cell_types : frozenset[str]
+        FrozenSet of valid cell types
+    cell_counts : dict[str, int], optional, default=None
+        Dictionary mapping cell type ontology term IDs (potentially suffixed) to the number of cells of that type.
+        This is used to exclude all descendants of a cell type if the cell type has the same number of cells as one
+        of its descendants. In these cases, the cell type is a redundant node and should be excluded.
+
+
+    Returns
+    -------
+    list[str] - An empty list indicates that the cell type is a redundant node and should be excluded.
+    """
+    # if the input cell types are suffixed, only consider descendants within the same group
+    if ";;" in cell_type:
+        prefix, suffix = cell_type.split(";;")
+        suffix = ";;" + suffix
+    else:
+        prefix = cell_type
+        suffix = ""
+
+    # find descendants of cell type and re-suffix them
+    relatives = [f"{i}{suffix}" for i in descendants(prefix)]
+    valid_relatives = list(valid_cell_types.intersection(relatives))
+    if cell_counts is not None:
+        relative_cell_counts = [
+            cell_counts.get(relative)
+            for relative in valid_relatives
```

Comment:
The indentation looks broken here (and it's missing a `:`)

Reply:
Haha, this is a multi-line list comprehension.